lib/Service/ImageService.php

```diff
@@ -0,0 +1,596 @@
+				'path' => $path,
+				'id' => $savedFile->getId(),
+				'textFileId' => $textFile->getId(),
+			];
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * Save an uploaded image in the attachment folder in a public context
+	 *
+	 * @param int|null $textFileId
+	 * @param string $newFileName
+	 * @param string $newFileContent
+	 * @param string $shareToken
+	 * @return array|string[]
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function uploadImagePublic(?int $textFileId, string $newFileName, string $newFileContent, string $shareToken): array {
+		if (!$this->hasUpdatePermissions($shareToken)) {
+			throw new Exception('No write permissions');
+		}
+		$textFile = $this->getTextFilePublic($textFileId, $shareToken);
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			$fileName = (string) time() . '-' . $newFileName;
+			$savedFile = $saveDir->newFile($fileName, $newFileContent);
+			$path = preg_replace('/^files/', '', $savedFile->getInternalPath());
+			return [
+				'name' => $fileName,
+				'path' => $path,
+				'id' => $savedFile->getId(),
+				'textFileId' => $textFile->getId(),
+			];
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * Copy a file from a user's storage in the attachment folder
+	 *
+	 * @param int $textFileId
+	 * @param string $path
+	 * @param string $userId
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 * @throws \OC\User\NoUserException
+	 */
+	public function insertImageFile(int $textFileId, string $path, string $userId): array {
+		$textFile = $this->getTextFile($textFileId, $userId);
+		if (!$textFile->isUpdateable()) {
+			throw new Exception('No write permissions');
+		}
+		$imageFile = $this->getFileFromPath($path, $userId);
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->copyImageFile($imageFile, $saveDir, $textFile);
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * @param File $imageFile
+	 * @param Folder $saveDir
+	 * @param File $textFile
+	 * @return array|string[]
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 */
+	private function copyImageFile(File $imageFile, Folder $saveDir, File $textFile): array {
+		$mimeType = $imageFile->getMimeType();
+		if (in_array($mimeType, ImageController::IMAGE_MIME_TYPES)) {
+			$fileName = (string) time() . '-' . $imageFile->getName();
+			$targetPath = $saveDir->getPath() . '/' . $fileName;
+			$targetFile = $imageFile->copy($targetPath);
+			$path = preg_replace('/^files/', '', $targetFile->getInternalPath());
```

Comment:
Might not work well with other storages than the user home (e.g. group folders)

I think something like this should also work to resolve the file within the user folder:

`$userFolder->getRelativePath($file->getPath())`

Reply:
Oh right. I'll keep that in mind. In this specific case, the `path` attribute is actually not used anymore in the frontend. It has been removed.